File: tests/kicad_xyce_plugin/simulation_parameters/step_parameters_test.py
from kicad_xyce_plugin.simulation_parameters import StepParameters


class TestStepParametersParsing:

    def test_parse_linear_step_implicit(self):
        # arrange
        directives = [".STEP R1 1k 10k 1k"]

        # act
        p = StepParameters.from_xyce_directives(directives)

        # assert
        assert p.enabled is True
        assert p.sweep_mode == "LIN"
        assert p.variable == "R1"
        assert p.start == "1k"
        assert p.stop == "10k"
        assert p.step == "1k"

    def test_parse_linear_step_explicit(self):
        # arrange
        directives = [".STEP LIN R1 1k 10k 1k"]

        # act
        p = StepParameters.from_xyce_directives(directives)

        # assert
        assert p.enabled is True
        assert p.sweep_mode == "LIN"
        assert p.variable == "R1"
        assert p.start == "1k"
        assert p.stop == "10k"
        assert p.step == "1k"

    def test_parse_dec_step(self):
        # arrange
        directives = [".STEP DEC R1 100 1MEG 10"]

        # act
        p = StepParameters.from_xyce_directives(directives)

        # assert
        assert p.enabled is True
        assert p.sweep_mode == "DEC"
        assert p.variable == "R1"
        assert p.start == "100"
        assert p.stop == "1MEG"
        assert p.points == "10"

    def test_parse_list_step(self):
        # arrange
        directives = [".STEP R1 LIST 1k 2k 5k"]

        # act
        p = StepParameters.from_xyce_directives(directives)

        # assert
        assert p.enabled is True
        assert p.sweep_mode == "LIST"
        assert p.variable == "R1"
        assert p.list_values == ("1k", "2k", "5k")

    def test_parse_data_step(self):
        # arrange
        directives = [".STEP DATA=myTable"]

        # act
        p = StepParameters.from_xyce_directives(directives)

        # assert
        assert p.enabled is True
        assert p.sweep_mode == "DATA"
        assert p.data_table_name == "myTable"

    def test_parse_data_step_with_spaces(self):
        # arrange
        directives = [".STEP DATA = myTable"]

        # act
        p = StepParameters.from_xyce_directives(directives)

        # assert
        assert p.enabled is True
        assert p.sweep_mode == "DATA"
        assert p.data_table_name == "myTable"

    def test_parsing_returns_disabled_when_no_directive(self):
        # arrange
        directives = [".TRAN 1u 1m"]

        # act
        p = StepParameters.from_xyce_directives(directives)

        # assert
        assert p.enabled is False

    def test_all_from_xyce_directives_returns_multiple_steps(self):
        # arrange - nested loops: inner first, outer last
        directives = [
            ".TRAN 1u 1m",
            ".STEP R1 1k 10k 1k",
            ".STEP TEMP 0 100 10",
        ]

        # act
        steps = StepParameters.all_from_xyce_directives(directives)

        # assert
        assert len(steps) == 2
        assert steps[0].variable == "R1"
        assert steps[1].variable == "TEMP"

    def test_all_from_xyce_directives_returns_empty_tuple_when_no_step(self):
        # arrange
        directives = [".TRAN 1u 1m"]

        # act
        steps = StepParameters.all_from_xyce_directives(directives)

        # assert
        assert steps == ()

    def test_all_from_xyce_directives_preserves_order(self):
        # arrange - three steps in declaration order
        directives = [
            ".STEP DEC R1 1 1k 5",
            ".STEP R2 LIST 100 200 300",
            ".STEP TEMP 0 100 25",
        ]

        # act
        steps = StepParameters.all_from_xyce_directives(directives)

        # assert
        assert len(steps) == 3
        assert steps[0].sweep_mode == "DEC"
        assert steps[1].sweep_mode == "LIST"
        assert steps[2].sweep_mode == "LIN"


class TestStepParametersGeneration:

    def test_generates_linear_step_with_explicit_lin_keyword(self):
        # arrange
        p = StepParameters(sweep_mode="LIN", variable="R1", start="1k", stop="10k", step="1k", enabled=True)

        # act
        directives = p.to_xyce_directives()

        # assert - LIN keyword is always explicit per project convention
        assert directives == [".STEP LIN R1 1k 10k 1k"]

    def test_generates_dec_step(self):
        # arrange
        p = StepParameters(sweep_mode="DEC", variable="R1", start="100", stop="1MEG", points="10", enabled=True)

        # act
        directives = p.to_xyce_directives()

        # assert
        assert directives == [".STEP DEC R1 100 1MEG 10"]

    def test_generates_list_step(self):
        # arrange
        p = StepParameters(sweep_mode="LIST", variable="R1", list_values=("1k", "2k", "5k"), enabled=True)

        # act
        directives = p.to_xyce_directives()

        # assert
        assert directives == [".STEP R1 LIST 1k 2k 5k"]

    def test_generates_data_step(self):
        # arrange
        p = StepParameters(sweep_mode="DATA", data_table_name="myTable", enabled=True)

        # act
        directives = p.to_xyce_directives()

        # assert
        assert directives == [".STEP DATA=myTable"]

    def test_generates_nothing_when_disabled(self):
        # arrange
        p = StepParameters(enabled=False)

        # act
        directives = p.to_xyce_directives()

        # assert
        assert directives == []

File: tests/simulation_parameters/step_parameters.test.cpp
#include <gtest/gtest.h>

#include "simulation_parameters/step_parameters.h"

// ========================================================================================
// from_single_directive
// ========================================================================================

TEST(StepParametersChecks, parse_linear_step_implicit) {
    // arrange
    const std::string directive = ".STEP R1 1k 10k 1k";
    // act
    const auto result = StepParameters::from_single_directive(directive);
    // assert
    ASSERT_TRUE(result.has_value());
    ASSERT_TRUE(result->enabled);
    ASSERT_EQ(result->sweep_mode, "LIN");
    ASSERT_EQ(result->variable, "R1");
    ASSERT_EQ(result->start, "1k");
    ASSERT_EQ(result->stop, "10k");
    ASSERT_EQ(result->step, "1k");
}

TEST(StepParametersChecks, parse_linear_step_explicit) {
    // arrange
    const std::string directive = ".STEP LIN R1 1k 10k 1k";
    // act
    const auto result = StepParameters::from_single_directive(directive);
    // assert
    ASSERT_TRUE(result.has_value());
    ASSERT_TRUE(result->enabled);
    ASSERT_EQ(result->sweep_mode, "LIN");
    ASSERT_EQ(result->variable, "R1");
    ASSERT_EQ(result->start, "1k");
    ASSERT_EQ(result->stop, "10k");
    ASSERT_EQ(result->step, "1k");
}

TEST(StepParametersChecks, parse_dec_step) {
    // arrange
    const std::string directive = ".STEP DEC R1 100 1MEG 10";
    // act
    const auto result = StepParameters::from_single_directive(directive);
    // assert
    ASSERT_TRUE(result.has_value());
    ASSERT_TRUE(result->enabled);
    ASSERT_EQ(result->sweep_mode, "DEC");
    ASSERT_EQ(result->variable, "R1");
    ASSERT_EQ(result->start, "100");
    ASSERT_EQ(result->stop, "1MEG");
    ASSERT_EQ(result->points, "10");
}

TEST(StepParametersChecks, parse_list_step) {
    // arrange
    const std::string directive = ".STEP R1 LIST 1k 2k 5k";
    // act
    const auto result = StepParameters::from_single_directive(directive);
    // assert
    ASSERT_TRUE(result.has_value());
    ASSERT_TRUE(result->enabled);
    ASSERT_EQ(result->sweep_mode, "LIST");
    ASSERT_EQ(result->variable, "R1");
    ASSERT_EQ(result->list_values.size(), 3);
    ASSERT_EQ(result->list_values[0], "1k");
    ASSERT_EQ(result->list_values[1], "2k");
    ASSERT_EQ(result->list_values[2], "5k");
}

TEST(StepParametersChecks, parse_data_step) {
    // arrange
    const std::string directive = ".STEP DATA=myTable";
    // act
    const auto result = StepParameters::from_single_directive(directive);
    // assert
    ASSERT_TRUE(result.has_value());
    ASSERT_TRUE(result->enabled);
    ASSERT_EQ(result->sweep_mode, "DATA");
    ASSERT_EQ(result->data_table_name, "myTable");
}

TEST(StepParametersChecks, parse_data_step_with_spaces) {
    // arrange
    const std::string directive = ".STEP DATA = myTable";
    // act
    const auto result = StepParameters::from_single_directive(directive);
    // assert
    ASSERT_TRUE(result.has_value());
    ASSERT_TRUE(result->enabled);
    ASSERT_EQ(result->sweep_mode, "DATA");
    ASSERT_EQ(result->data_table_name, "myTable");
}

TEST(StepParametersChecks, parsing_returns_disabled_when_no_directive) {
    // arrange
    const std::string directive = ".TRAN 1u 1m";
    // act
    const auto result = StepParameters::from_single_directive(directive);
    // assert
    ASSERT_FALSE(result.has_value());
}

TEST(StepParametersChecks, from_xyce_directives_returns_multiple_steps) {
    // arrange - nested loops: inner first, outer last
    const std::vector<std::string> directives = {
        ".STEP R1 1k 10k 1k",
        ".STEP C1 1n 100n 10",
    };
    // act
    const auto result = StepParameters::from_xyce_directives(directives);
    // assert
    ASSERT_TRUE(result.has_value());
    ASSERT_TRUE(result->enabled);
    ASSERT_EQ(result->sweep_mode, "LIN");
    ASSERT_EQ(result->variable, "R1");
    ASSERT_EQ(result->start, "1k");
    ASSERT_EQ(result->stop, "10k");
    ASSERT_EQ(result->step, "1k");
}

TEST(StepParametersChecks, from_xyce_directives_empty_returns_disabled) {
    // arrange
    const std::vector<std::string> directives = {};
    // act
    const auto result = StepParameters::from_xyce_directives(directives);
    // assert
    ASSERT_TRUE(result.has_value());
    ASSERT_FALSE(result->enabled);
}

TEST(StepParametersChecks, from_xyce_directives_non_step_returns_disabled) {
    // arrange
    const std::vector<std::string> directives = {".TRAN 1u 1m"};
    // act
    const auto result = StepParameters::from_xyce_directives(directives);
    // assert
    ASSERT_TRUE(result.has_value());
    ASSERT_FALSE(result->enabled);
}

// ========================================================================================
// all_from_xyce_directives
// ========================================================================================

TEST(StepParametersChecks, all_from_xyce_directives_returns_multiple_steps) {
    // arrange
    const std::vector<std::string> directives = {
        ".STEP R1 1k 10k 1k",
        ".STEP C1 1n 100n 10n",
    };
    // act
    const auto results = StepParameters::all_from_xyce_directives(directives);
    // assert
    ASSERT_EQ(results.size(), 2);
    ASSERT_TRUE(results[0].enabled);
    ASSERT_EQ(results[0].variable, "R1");
    ASSERT_TRUE(results[1].enabled);
    ASSERT_EQ(results[1].variable, "C1");
}

// ========================================================================================
// from_xyce_directives
// ========================================================================================

TEST(StepParametersChecks, from_xyce_directives_returns_first_step) {
    // arrange
    const std::vector<std::string> directives = {
        ".STEP R1 1k 10k 1k",
        ".STEP C1 1n 100n 10n",
    };
    // act
    const auto result = StepParameters::from_xyce_directives(directives);
    // assert
    ASSERT_TRUE(result->enabled);
    ASSERT_EQ(result->variable, "R1");
}

TEST(StepParametersChecks, from_xyce_directives_disabled_when_no_step) {
    // arrange
    const std::vector<std::string> directives = {
        ".TRAN 1u 1m",
    };
    // act
    const auto result = StepParameters::from_xyce_directives(directives);
    // assert
    ASSERT_FALSE(result->enabled);
}

// ========================================================================================
// to_xyce_directives
// ========================================================================================

TEST(StepParametersChecks, generates_linear_step_directive) {
    // arrange
    const StepParameters params("LIN", "R1", "1k", "10k", "1k", "", {}, "", true);
    // act
    const auto directives = params.to_xyce_directives();
    // assert
    ASSERT_EQ(directives.size(), 1);
    ASSERT_EQ(directives[0], ".STEP LIN R1 1k 10k 1k");
}

TEST(StepParametersChecks, generates_dec_step_directive) {
    // arrange
    const StepParameters params("DEC", "R1", "100", "1MEG", "", "10", {}, "", true);
    // act
    const auto directives = params.to_xyce_directives();
    // assert
    ASSERT_EQ(directives.size(), 1);
    ASSERT_EQ(directives[0], ".STEP DEC R1 100 1MEG 10");
}

TEST(StepParametersChecks, generates_list_step_directive) {
    // arrange
    const StepParameters params("LIST", "R1", "", "", "", "", {"1k", "2k", "5k"}, "", true);
    // act
    const auto directives = params.to_xyce_directives();
    // assert
    ASSERT_EQ(directives.size(), 1);
    ASSERT_EQ(directives[0], ".STEP R1 LIST 1k 2k 5k");
}

TEST(StepParametersChecks, generates_data_step_directive) {
    // arrange
    const StepParameters params("DATA", "", "", "", "", "", std::vector<std::string>{}, "myTable", true);
    // act
    const auto directives = params.to_xyce_directives();
    // assert
    ASSERT_EQ(directives.size(), 1);
    ASSERT_EQ(directives[0], ".STEP DATA=myTable");
}

TEST(StepParametersChecks, generates_disabled_step_returns_empty) {
    // arrange
    const StepParameters params("", "", "", "", "", "", std::vector<std::string>{}, "", false);
    // act
    const auto directives = params.to_xyce_directives();
    // assert
    ASSERT_EQ(directives.size(), 0);
}

// ...existing code...
